File: app/src/main/java/com/example/macarus0/walkiewalkie/view/DeletablePhotoListAdapter.java
package com.example.macarus0.walkiewalkie.view;

import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.macarus0.walkiewalkie.R;
import com.example.macarus0.walkiewalkie.data.PhotoItem;
import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class DeletablePhotoListAdapter extends RecyclerView.Adapter<DeletablePhotoListAdapter.ViewHolder> {

    private static final String TAG = "DeletablePhotoListAdapter";
    List<? extends PhotoItem> mPhotos;
    DeleteHandler mDeleteHandler = null;
    boolean showPhotoLabels = false;

    public void setPlaceholderImage(Drawable placeholderImage) {
        this.placeholderImage = placeholderImage;
    }

    Drawable placeholderImage;

    public void setDeleteHandler(DeleteHandler mDeleteHandler) {
        this.mDeleteHandler = mDeleteHandler;
    }

    public void setShowPhotoLabels(boolean showPhotoLabels) {
        this.showPhotoLabels = showPhotoLabels;
    }

    public void setPhotos(List<? extends PhotoItem> photos) {
        mPhotos = photos;
        Log.i(TAG, "setPhotos: Loaded photos " + photos.size());
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public DeletablePhotoListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.contact_card, parent, false);
        return new DeletablePhotoListAdapter.ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull DeletablePhotoListAdapter.ViewHolder holder, int position) {
        PhotoItem photoItem = this.mPhotos.get(position);
        Log.i(TAG, "onBindViewHolder: Loading image " + photoItem.getPhotoUri());
        Picasso.get().load(photoItem.getPhotoUri()).placeholder(this.placeholderImage).into(holder.walkImage);
        if(showPhotoLabels) {
            holder.tv.setText(photoItem.getName());
        }
        holder.removeButton.setOnClickListener((View v) ->
                mDeleteHandler.onDeletePress(photoItem.getId()));

    }

    @Override
    public int getItemCount() {
        if (mPhotos == null) {
            Log.i(TAG, "getItemCount: No Photos Yet");
            return 0;
        }
        Log.i(TAG, "getItemCount: photoCount " + mPhotos.size());
        return mPhotos.size();
    }

    interface DeleteHandler {
        void onDeletePress(long id);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.contact_name)
        TextView tv;
        @BindView(R.id.item_remove)
        AppCompatImageButton removeButton;

        @BindView(R.id.contact_image)
        ImageView walkImage;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            removeButton.setVisibility(View.VISIBLE);

            if (!showPhotoLabels) {
                tv.setVisibility(View.INVISIBLE);
            }
        }

    }

}

File: app/src/main/java/com/example/macarus0/walkiewalkie/view/DeletablePhotoListAdapter.kt
package com.example.macarus0.walkiewalkie.view

import android.graphics.drawable.Drawable
import androidx.appcompat.widget.AppCompatImageButton
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.example.macarus0.walkiewalkie.R
import com.example.macarus0.walkiewalkie.data.PhotoItem
import com.squareup.picasso.Picasso

import butterknife.BindView
import butterknife.ButterKnife

class DeletablePhotoListAdapter : RecyclerView.Adapter<DeletablePhotoListAdapter.ViewHolder>() {
    internal var photos: List<PhotoItem>? = null
        set(value) {
            field = value
            Log.i(TAG, "setPhotos: Loaded photos " + photos?.size)
            notifyDataSetChanged()
        }

    lateinit var deleteHandler: (Long) -> Unit
    var showPhotoLabels = false
    lateinit var placeholderImage: Drawable

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeletablePhotoListAdapter.ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.contact_card, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: DeletablePhotoListAdapter.ViewHolder, position: Int) {
        val photoItem = this.photos!![position]
        Log.i(TAG, "onBindViewHolder: Loading image " + photoItem.photoUri)
        Picasso.get().load(photoItem.photoUri).placeholder(this.placeholderImage).into(holder.walkImage)
        if (showPhotoLabels) {
            holder.tv!!.text = photoItem.name
        }
        holder.removeButton!!.setOnClickListener { _: View -> deleteHandler(photoItem.id) }

    }

    override fun getItemCount(): Int {
        if (photos == null) {
            Log.i(TAG, "getItemCount: No Photos Yet")
            return 0
        }
        Log.i(TAG, "getItemCount: photoCount " + photos!!.size)
        return photos!!.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.contact_name)
        @JvmField  var tv: TextView? = null
        @BindView(R.id.item_remove)
        @JvmField  var removeButton: AppCompatImageButton? = null

        @BindView(R.id.contact_image)
        @JvmField  var walkImage: ImageView? = null

        init {
            ButterKnife.bind(this, itemView)
            removeButton!!.visibility = View.VISIBLE

            if (!showPhotoLabels) {
                tv!!.visibility = View.INVISIBLE
            }
        }

    }

    companion object {

        private const val TAG = "DeletablePhotoListAdapter"
    }

}
